use-cases/eurac/hpo.py

```diff
@@ -0,0 +1,202 @@
+        ray.init(
+            address=os.environ["ip_head"],
+            _node_ip_address=os.environ["head_node_ip"],
+        )
+
+        # Define the search space for hyperparameters
+        search_space = {
+            'batch_size': tune.choice([64, 128, 256]),
+            'lr': tune.uniform(1e-5, 1e-3)
+        }
+
+        # TuneConfig for configuring search algorithm and scheduler
+        tune_config = tune.TuneConfig(
+            metric=args.metric,  # Metric to optimize (loss by default)
+            mode="min",  # Minimize the loss
+            num_samples=args.num_samples  # Number of trials to run
+        )
+
+        # Ray's RunConfig for experiment name and stopping criteria
+        run_config = train.RunConfig(
+            name="Eurac-Ray-Experiment",
+            stop={"training_iteration": args.max_iterations}
+        )
+
+        # Determine GPU and CPU utilization per trial
+        # We are allocating all available ressources per node evenly across trials
+        ngpus_per_trial = max(1, args.ngpus // args.num_samples)
+        ncpus_per_trial = max(1, args.ncpus // args.num_samples)
+
+        # Set resource allocation for each trial (number of GPUs and/or number of CPUs)
+        resources_per_trial = {"gpu": ngpus_per_trial, "cpu": ncpus_per_trial}
+
+        # Set up Ray Tune Tuner
+        tuner = tune.Tuner(
+            tune.with_resources(run_trial, resources=resources_per_trial),
+            tune_config=tune_config,
+            run_config=run_config,
+            param_space=search_space  # Search space defined above
+        )
+
+        # Run the hyperparameter optimization and get results
+        result_grid = tuner.fit()
+
+    else:
+        # Load results from an earlier Ray Tune run
+        print(f"Loading results from {args.experiment_path}...")
+
+        # Restore tuner from saved results
+        restored_tuner = tune.Tuner.restore(
+            args.experiment_path,
+            trainable=run_trial
+        )
+        result_grid = restored_tuner.get_results()
+
+    # Display experiment statistics
+    print(f"Number of errored trials: {result_grid.num_errors}")
+    print(f"Number of terminated trials: {result_grid.num_terminated}")
+    print(f"Ray Tune experiment path: {result_grid.experiment_path}")
+
+    # Get the best result based on the last 10 iterations' average
+    best_result = result_grid.get_best_result(
+        scope="last-10-avg",
+        metric=args.metric,
+        mode="min"
+    )
+    print(f"Best result: {best_result}")
+
+    # Print a dataframe with all trial results
+    result_df = result_grid.get_dataframe()
+    print(f"All results dataframe: {result_df}")
+    print(f"All result columns: {result_df.columns}")
+
+    # Plot the results for all trials
+    plot_results(result_grid, metric=args.metric, filename="ray-loss-plot.png")
+    plot_results(result_grid, metric="train_loss",
```

Comment:
Here you have two args on the first line and one arg on the second line which is not very readable. I would suggest either doing 

plot_results(
    arg1, arg2, arg2
)

or

plot_results(
    arg1, 
    arg2, 
    etc.
)

This offers quite a bit better readability IMO